src/main/scala/org/clulab/wm/eidos/context/TimeNormFinder.scala

```diff
@@ -30,149 +43,187 @@ object TimeNormFinder {
+        //assert(end2 <= end3 && end3 <= sentenceText.length)
+
+        //assert(start3 <= end3)
+        // yield the context interval adjusting the span to full-text character offsets
+        TextInterval(sentenceStart + start3, sentenceStart + end3)
       }
-      val (contextSentenceIndexes: Array[Int], contextSpans: Array[(Int, Int)]) = sentenceContexts.flatten.unzip
-
-      // run the timenorm parser over batches of contexts to find and normalize time expressions
-      val contextTimeExpressions: Array[Array[TimeExpression]] = eidosDocument.dctString match {
-        case Some(dctString) =>
-          val Array(dct: TimExInterval) = parser.parse(dctString)
-          // use a SimpleInterval (character span = None) so it doesn't interfere with character span of TimeExpressions
-          val dctSimpleInterval = SimpleInterval(dct.start, dct.end)
-          eidosDocument.dct = Some(DCT(dctSimpleInterval, dctString))
-          contextSpans.sliding(BATCH_SIZE, BATCH_SIZE).
-            flatMap(batch => parseBatch(text, batch, textCreationTime = dctSimpleInterval)).toArray
-        case None =>
-          contextSpans.sliding(BATCH_SIZE, BATCH_SIZE).
-            flatMap(batch => parseBatch(text, batch)).toArray
+
+      // merge overlapping contexts
+      val mergedIntervals = singleIntervals.sorted.foldLeft(Seq.empty[TextInterval]) {
+        case (Seq(), context) => Seq(context)
+        case (init :+ last, context) if last.end > context.start => init :+ TextInterval(last.start, context.end)
+        case (contexts, context) => contexts :+ context
       }
 
-      // create mentions for each of the time expressions that were found
-      for {
-        (sentenceIndex, timeExpressions) <- contextSentenceIndexes zip contextTimeExpressions
-        timeExpression <- timeExpressions
-      } yield {
-        val sentence = eidosDocument.sentences(sentenceIndex)
-        val Some((timeTextStart, timeTextEnd)) = timeExpression.charSpan
-        val timeTextInterval = Interval(timeTextStart, timeTextEnd)
-        val timeText = text.substring(timeTextStart, timeTextEnd)
-
-        // find the words covered by the time expression (only needed because TextBoundMention requires it)
-        val wordIndices = sentence.words.indices.filter { i =>
-          sentence.startOffsets(i) < timeTextEnd && timeTextStart < sentence.endOffsets(i)
-        }
+      // get the text for each context
+      for (interval <- mergedIntervals) yield {
+        (sentenceIndex, (interval.start, interval.end))
+      }
+    }
+    val (contextSentenceIndexes: Array[Int], contextSpans: Array[(Int, Int)]) = sentenceContexts.flatten.unzip
+
+    // run the timenorm parser over batches of contexts to find and normalize time expressions
+    val dctOpt = DctDocumentAttachment.getDct(doc)
+    val contextTimeExpressions: Array[Array[TimeExpression]] = dctOpt match {
+      case Some(dct: DCT) =>
+        val simpleInterval = dct.interval.asInstanceOf[SimpleInterval]
```

Comment:
This cast should not be needed.

Reply:
Thanks!  That's a remnant of the old code.  Gone.